torch/csrc/autograd/function.h

```diff
@@ -521,6 +521,11 @@ struct TORCH_API Node : std::enable_shared_from_this<Node> {
     return retains_grad_hooks_;
   }
 
+  virtual void set_non_differentiable_idx(
```

Comment:
This feels not right.
I feel like this shouldn't be on the top level Node class.
I like the idea of doing this setting in the shared code but since you need special handling in PyNode and CppNode anyways, I feel like the apply code for each of these could update these fields directly.

Reply:
Yeah thats fair, let me try something else.

Reply:
Using a new approach now that still does things in the shared code, but no longer places methods on the top level Node class.
The new sin we're committing here though is that we're overloading the default constructed InputMetadata to mean something special for custom functions, but maybe it isn't that bad.